src/compiler/checker.ts

```diff
@@ -14313,20 +14322,25 @@ namespace ts {
             if (!(isMatchingSignature(source, target, partialMatch))) {
                 return Ternary.False;
             }
-            // Check that the two signatures have the same number of type parameters. We might consider
-            // also checking that any type parameter constraints match, but that would require instantiating
-            // the constraints with a common set of type arguments to get relatable entities in places where
-            // type parameters occur in the constraints. The complexity of doing that doesn't seem worthwhile,
-            // particularly as we're comparing erased versions of the signatures below.
+            // Check that the two signatures have the same number of type parameters.
             if (length(source.typeParameters) !== length(target.typeParameters)) {
                 return Ternary.False;
             }
-            // Spec 1.0 Section 3.8.3 & 3.8.4:
-            // M and N (the signatures) are instantiated using type Any as the type argument for all type parameters declared by M and N
-            source = getErasedSignature(source);
-            target = getErasedSignature(target);
+            // Check that type parameter constraints and defaults match. If they do, instantiate the source
+            // signature with the type parameters of the target signature and continue the comparison.
+            if (target.typeParameters) {
+                const mapper = createTypeMapper(source.typeParameters!, target.typeParameters);
+                for (let i = 0; i < target.typeParameters.length; i++) {
+                    const s = source.typeParameters![i];
+                    const t = target.typeParameters[i];
+                    if (!(s === t || compareTypes(instantiateType(getConstraintFromTypeParameter(s), mapper) || unknownType, getConstraintFromTypeParameter(t) || unknownType) &&
+                        compareTypes(instantiateType(getDefaultFromTypeParameter(s), mapper) || unknownType, getDefaultFromTypeParameter(t) || unknownType))) {

```

Comment:
Yeah, I don't think we can consider defaults when checking identity - if we do, then two conditional extends clauses that only differ in a local type parameter default (which doesn't affect relationship checking) will fail to be seen as identical.

Reply:
Eg, `(A extends <T = {}>() => T ? 1 : 0) extends (A extends <T = unknown>() => T ? 1 : 0) ? true : false` should resolve to `true`, I think, not `false`.

Reply:
I'm not sure I understand how that example is relevant. Best I can tell at no point will we compare signatures for identity in the example.

Reply:
Conditional `extends` clauses are compared using `isTypeIdenticalTo` when checking if one conditional extends the other in the above example. Apparently this is already in use in the wild as a way to observe type "identity" in complex type toolbelt libraries.

Reply:
Ok, I see that logic in `structuredTypeRelatedTo`. Still, as far as I'm concerned, those two types aren't identical and I can't think of a meaningful definition of "identical" that would make it true for them, but not for the other cases we care about. It would just seem really odd to make an exception for local type parameter defaults, but nothing else.

Reply:
I mean, we let you _assign_ a function with a default to a function with a different default and vice-versa - it's just the typesystem equivalent of aliasing. Generally defaults aren't observable up until the point where inference fails - this exposes them into relationship checking before inference is ever invoked, success or no.

Reply:
I still don't think that merits a special exception to identity checking. The identity relation should be concerned with accurately identifying semantically identical types, and it does a better job of that now. I think we need to get this merged.